lib/logger/lib/logger.ex

```diff
@@ -695,49 +680,58 @@ defmodule Logger do
           :ok | {:error, :noproc} | {:error, term}
   def bare_log(level, chardata_or_fun, metadata \\ []) do
     case __should_log__(level) do
-      :error -> :ok
-      info -> __do_log__(info, chardata_or_fun, metadata)
+      nil -> :ok
+      level -> __do_log__(level, chardata_or_fun, Map.new(metadata))
     end
   end
 
   @doc false
+  def __should_log__(:warn) do
+    # TODO: Uncomment this
+    # bare_log(:warning, "`:warn` level is deprecated, you should use `:warning`")
+
+    __should_log__(:warning)
+  end
+
   def __should_log__(level) when level in @levels do
-    case __metadata__() do
-      {true, pdict} ->
-        case Logger.Config.log_data(level) do
-          {:discard, _config} -> :error
-          {mode, config} -> {level, mode, config, pdict}
-        end
+    %{level: primary_level} = :logger.get_primary_config()
 
-      {false, _} ->
-        :error
+    if enabled?(self()) and compare_levels(level, primary_level) != :lt do
+      level
     end
   end
 
-  @doc false
-  def __do_log__({level, mode, config, pdict}, chardata_or_fun, metadata)
-      when is_list(metadata) do
-    %{utc_log: utc_log?, truncate: truncate} = config
-    metadata = [pid: self()] ++ into_metadata(metadata, pdict)
-
-    case normalize_message(chardata_or_fun, metadata) do
-      {message, metadata} ->
-        tuple = {Logger, truncate(message, truncate), Logger.Utils.timestamp(utc_log?), metadata}
-
-        try do
-          notify(mode, {level, Process.group_leader(), tuple})
-          :ok
-        rescue
-          ArgumentError -> {:error, :noproc}
-        catch
-          :exit, reason -> {:error, reason}
-        end
-
+  def __do_log__(level, fun, metadata) when is_function(fun, 0) and is_map(metadata) do
+    case fun.() do
       :skip ->
         :ok
+
+      {msg, meta} ->
+        :logger.log(level, msg, Enum.into(meta, add_elixir_domain(metadata)))
+
+      msg when is_binary(msg) ->
+        :logger.log(level, msg, add_elixir_domain(metadata))
     end
   end
 
+  def __do_log__(level, chardata, metadata)
+      when (is_binary(chardata) or is_list(chardata)) and is_map(metadata) do
+    :logger.log(level, chardata, add_elixir_domain(metadata))
+  end
+
+  # # TODO: Remove that in Elixir 2.0
+  def __do_log__(level, atom, metadata) do
+    # TODO: Uncomment this
+    bare_log(:warning, "Passing #{inspect(other)} to Logger is deprecated, expected a binary or an iolist")
```

Comment:
```suggestion
    IO.warn("Passing #{inspect(other)} to Logger is deprecated, expected a binary or an iolist")
```

We generally do not emit deprecation warnings through Logger. Also, please remove the TODO above.

Reply:
Also note you have to rename the variable atom to other, above and below. :)